sql/catalog/descs: combine synthetic descriptors with stored descriptors

postamar opened the pull request:
This allows synthetic descriptors to show up in virtual tables.

Release justification: part of bug fix.

Release note: None

ajwerner committed:
sql/catalog/descs: combine synthetic descriptors with stored descriptors

This allows synthetic descriptors to show up in virtual tables.

Release justification: part of bug fix.

Release note: None

ajwerner committed:
sql: rework some internal executor logic related to synthetic descripotrs

The pre-existing logic would reset the synthetic descriptors whenever anything
was run on behalf of the internal planner. That's never what we want. We
prevent that by attaching the logic to reset the synthetic descriptors to
whether the internal executor has attached synthetic descriptors, which
it never has when using the planner's state.

Release justification: part of bug fix

Release note: None

Commit:
descs,catkv: rewrite Collection backend

This rewrite was motivated by the upcoming need to validate descriptors
up to different levels depending on why they were being read. Doing so
requires tracking the validation level in a more granular fashion within
the Collection's caches. Trying to make this work within the existing
scheme where the uncommitted and stored descriptors are all in the same
layer turned out to be painful and hard to reason about.

This commit does several things:
1. The uncommitted descriptors are spun off into their own distinct layer
   again.
2. The stored descriptors now mirror what's in storage and so that whole
   layer could effectively be moved to catkv.
3. Execution flow through the layers in the Collection is now
   straightforwardly implemented and easy to reason about. Validation and
   hydration now take place in unique, well-defined steps at the very end
   of a lookup, be it by name or by ID.
3. Instances of this new stored descriptors layer are now also used by
   the lease manager, the Direct() interface provided by the Collection,
   etc.
4. As a result the logic in catkv could be significantly cleaned up.
   In particular, lookups for descriptor validation are significantly
   less convoluted now.
5. The SystemNamespaceCache object was also moved to catkv, renamed to
   SystemDatabaseCache and now caches based on the cluster version.

Despite significant rewriting of the inner workings of the descriptors
collection, this commit should not functionally change anything.

Informs #85263.

Release justification: low-risk, high-benefit change.
Release note: None

Commit:
backupinfo: fix bug in writing names to metadata

Recent changes to the descriptors collection have uncovered some
brittleness in the logic for writing namespace entries for dropped
descriptors when creating backups.

This commit makes it:
- write deletion entries for descriptors other than tables, preventing
  possible catalog corruption.
- throw a better error message when attempting to write two identical
  keys consecutively.

Release justification: low-risk change
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/87144)
<!-- Reviewable:end -->

Commit:
nstree: rename Map to NameMap, add and adopt IDMap

This commit adds a new nstree.Map-like data structure without any
by-name operations.

Release justification: low-risk change which does not change
functionality
Release note: None

postamar commented:
Closed in favour of https://github.com/cockroachdb/cockroach/pull/87067, whose branch has been fast-forwarded to this one.
